h/migrations/versions/7bf167611bc3_add_the_groupscope_table.py

```diff
@@ -0,0 +1,39 @@
+"""
+Add the GroupScope table.
+
+Revision ID: 7bf167611bc3
+Revises: c943c3f8a7e5
+Create Date: 2018-02-08 11:00:50.420618
+"""
+
+from __future__ import unicode_literals
+
+from alembic import op
+import sqlalchemy as sa
+
+revision = '7bf167611bc3'
+down_revision = 'c943c3f8a7e5'
+
+
+def upgrade():
+    op.create_table(
+        'groupscope',
+        sa.Column('id', sa.Integer, autoincrement=True, primary_key=True),
+        sa.Column('hostname', sa.UnicodeText, nullable=False, unique=True),
```

Comment:
Does `unique=True` here make this column unique across _all_ groups, not just unique _within a group_. In other words, you want a unique constraint on `(id, hostname)` tuples.

Reply:
My inclination is that we should make this a `host` (optionally including port number) rather than `hostname`. Using a port number isn't something I would expect to use widely but I can imagine it would be useful for local testing. It would also make our concept of what is included in a host consistent with a web origin, and I have a hunch that aligning those two security boundaries could be useful.

Reply:
`unique=True` does indeed mean that the same hostname can't be entered into the `groupscope` table twice at all. The way to have two groups related to the same hostname would be a 1-many relationship from one `groupscope` row to many `group` table rows. I'm not sure that we'll want hostname to remain globablly unique in the `groupscope` table forever, but went with it for now.

Reply:
Where does this definition of host come from? I don't _think_ the port is part of the host in rfc3986. If we want it to be consistent with a web origin (which does sound like a good idea to me) I wonder if we should just call it `origin`?

Reply:
In Python 2.7's [urlparse module](https://docs.python.org/2.7/library/urlparse.html) an origin (hostname including subdomains, plus port) is called a "netloc":

```python
> urlparse.urlparse('https://foo.example.com:5000/path/to/page?q=yesd')
ParseResult(scheme='https', netloc='foo.example.com:5000', path='/path/to/page', params='', query='q=yesd', fragment='')
```

Reply:
Rob mentioned in a [comment elsewhere](https://github.com/hypothesis/h/pull/4778#issuecomment-364189062) that "host" (meaning hostname + port) comes from [the URL Web API](https://developer.mozilla.org/en-US/docs/Web/API/URL) (in that API hostname is just the hostname whereas host is hostname + port).

I think we have 3 options at this point:

1. host
2. origin
3. netloc

I think any one of those three is fine

Reply:
Ah, an origin also includes the scheme, so it's not actually the same as host or netloc

Reply:
> unique=True does indeed mean that the same hostname can't be entered into the groupscope table twice at all.

Oh right sorry, I missed the fact that the group ID doesn't appear in this table because of course this is an N:M relation. 🤦‍♂️ 

> I wonder if we should just call it origin

"origin" is a well-defined concept on the web that includes protocol + hostname + port, so I would be in favour of this. That does raise a question of whether we include the protocol and if so do we do HTTP/HTTPS normalization. We might find that it is too HTTP-specific, all of our use cases right now involve HTTP URIs.

>. In Python 2.7's urlparse module an origin (hostname including subdomains, plus port) is called a "netloc":

My impression was that "host" was used in multiple languages (at least Go, JS) and "netloc" was more Python-specific. I may be wrong about that.

Reply:
```
> urlparse.urlparse('https://foo.example.com:5000/path/to/page?q=yesd').netloc
'foo.example.com:5000'

> new URL("https://foo.example.com:5000/path/to/page?q=yesd").host
'foo.example.com:5000'

> new URL("https://foo.example.com:5000/path/to/page?q=yesd").origin
'https://foo.example.com:5000'
```

Reply:
I vote `origin`; `host` is OK; I don't like `netloc` because it is python-specific.

Reply:
I think having the scheme in there probably just introduces potential "why isn't my group appearing?" problems for no benefit (I don't think we actually need the scheme in there for security reasons). So I'm gonna vote for host which is what Rob originally suggested

Reply:
@lyzadanger do you think we should keep the scheme on there? So we scope a group to `https://nytimes.com` or `http://localhost:5324`?

Reply:
`host` doesn't typically include scheme, right? So, I'd vote no, if we're breaking up scope into `host` and other chunks...

Reply:
> I think having the scheme in there probably just introduces potential "why isn't my group appearing?" problems for no benefit 

Perhaps we could solve that by mandating HTTPS 😁 (or re-using the "httpx" normalization scheme that we use for document URIs).